eureka2-write-server/src/main/java/com/netflix/eureka2/server/service/EurekaWriteServerSelfRegistrationService.java

```diff
@@ -33,22 +41,36 @@ public void init() {
     }
 
     @Override
-    public void cleanUpResources() {
-        // no-op
-    }
-
-    @Override
-    public Observable<Void> register(final InstanceInfo instanceInfo) {
-        return registry.register(instanceInfo, selfSource).ignoreElements().cast(Void.class);
+    public Observable<Void> connect(final Observable<InstanceInfo> registrant) {
+        final AtomicReference<InstanceInfo> instanceInfoRef = new AtomicReference<>();
+        return registrant.replay(1).refCount()
+                .flatMap(new Func1<InstanceInfo, Observable<Void>>() {
+                    @Override
+                    public Observable<Void> call(InstanceInfo instanceInfo) {
+                        logger.info("registering self InstanceInfo {}", instanceInfo);
+                        instanceInfoRef.set(instanceInfo);
+                        return registry.register(instanceInfo, selfSource).ignoreElements().cast(Void.class);
+                    }
+                })
+                .doOnUnsubscribe(new Action0() {
+                    @Override
+                    public void call() {
+                        InstanceInfo info = instanceInfoRef.getAndSet(null);
+                        if (info != null) {
+                            logger.info("unregistering self InstanceInfo {}", info);
+                            registry.unregister(info, selfSource).subscribe();
+                        }
+                    }
+                });
     }
 
     @Override
-    public Observable<Void> unregister(InstanceInfo instanceInfo) {
-        return registry.unregister(instanceInfo, selfSource).ignoreElements().cast(Void.class);
+    public Source getSource() {
```

Comment:
I think at one time we may have queried the WriteSelfRegService for it's source, but looks like that dependency is gone now, so this can be removed.